client/client.go

```diff
@@ -1832,6 +1838,15 @@ func (c *Client) periodicSnapshot() {
 
 // run is a long lived goroutine used to run the client. Shutdown() stops it first
 func (c *Client) run() {
+
+	// Block until we've registered at least once so that we know the server has
+	// our node secret and we can authenticate
+	select {
+	case <-c.registeredCh:
+	case <-c.shutdownCh:
+		return
+	}
```

Comment:
This substantially changes what happens if a disconnected client agent with workloads restarts. Currently if a workload crashes, it can be restarted locally without reconnecting to the server. With this change _after a restart_ the client will be blocked here and not monitoring workloads for exits, so they will not be restarted.

I think this also breaks disconnected client agents with Consul based TTL health checks as Nomad won't heartbeat those checks until the runners start.

Luckily I think there's a pretty easy fix that's also a nice optimization: **persist whether first registration has occurred.** This means we only ever block on for first registration when there's no workloads running anyway! If there are runners running then we _know_ we've successfully registered so we should start running them again ASAP.

### Refactoring

This does make me realize I don't think anyone writing a Nomad Client from scratch would structure the code this way... breaking the currently monolithic `NewClient` into its 3 actual phases:

1. NewClient - initialize the struct, no side effects or IO
2. Register - only blocks on first run
3. Run - ...the runners! all the non-registration goroutines get spawned here

...would maybe make all of this easier to reason about, but that would be a pretty major refactor and touch innumerable tests. I highly doubt this is worth pursuing, but it's fun to dream.

Reply:
Gah, I totally missed the restart case :facepalm: Persisting registration is simple enough, will fix. And that'll probably make an easy way to block the RPCs specifically as well (as noted above).

I sort of wish the server was what handed out the node secret because that would make it really easy to identify a node that hadn't yet registered because it wouldn't yet have a secret. Maybe we can revisit that workflow when we look at secure introduction of nodes in the near-ish future, and that might be a good time to consider the refactoring you're talking about here as well.